### PR TITLE
feat: add `AddSubtractExpr`

### PR DESCRIPTION
**Pull request by iajoiner**
# Rationale for this change
We need to add support for arithmetic. Let's start from + and -.

# What changes are included in this PR?
- Add support for + and - in `proof-of-sql` crate
- generalize `scale_and_subtract_eval` to `scale_and_add_subtract_eval`
- make sure we pass in scales in any operation that can involve decimals

# Are these changes tested?
Yes.

### CODE REVIEW

**Review by JayWhite2357 — COMMENTED** [review]
Please add some tests that deal with overflow. I'm concerned that it isn't handled properly.

File: crates/proof-of-sql/src/base/database/column.rs [hunk 1]
```diff
@@ -240,6 +240,44 @@ impl ColumnType {
         )
     }
 
+    /// Returns the number of bits in the integer type if it is an integer type. Otherwise, return None.
+    pub fn to_integer_bits(&self) -> Option<usize> {
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/database/column.rs:244]: Not a big fan of this function. At the very least it should be private.

My preference would be for the `max_integer_type` function to just have a (admittedly tedious) match.
```suggestion
    fn to_integer_bits(&self) -> Option<usize> {
```

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    lhs: Column<'a, S>,
+    rhs: Column<'a, S>,
+    alloc: &'a Bump,
+    is_subtract: bool,
+) -> ConversionResult<Column<'a, S>> {
+    let lhs_len = lhs.len();
+    let rhs_len = rhs.len();
+    if lhs_len != rhs_len {
+        return Err(ConversionError::DifferentColumnLength(lhs_len, rhs_len));
+    }
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/numerical_util.rs:65]: I think this can safely be an `assert!`, in which case this doesn't need to return a `Result`.

The main reason is that this method should never fail since it is within the prover.

File: crates/proof-of-sql/src/sql/ast/numerical_util.rs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+/// The counterpart of `try_add_subtract_columns` for evaluating decimal expressions.
+pub(crate) fn scale_and_add_subtract_eval<S: Scalar>(
+    lhs_eval: S,
+    rhs_eval: S,
+    lhs_scale: i8,
+    rhs_scale: i8,
+    is_subtract: bool,
+) -> ConversionResult<S> {
+    let max_scale = lhs_scale.max(rhs_scale);
+    let scaled_lhs_eval = scale_scalar(lhs_eval, max_scale - lhs_scale)?;
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/numerical_util.rs:91]: We should be able to remove the `Result` from this function as well.
```suggestion
    let scaled_lhs_eval = scale_scalar(lhs_eval, max_scale - lhs_scale).expect("scaling factor should not be negative");
```

File: crates/proof-of-sql/src/sql/ast/add_subtract_expr_test.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/add_subtract_expr_test.rs:1]: I would like to see extensive overflow tests in this file.
Something to the effect of `SELECT A+B, A-B FROM table` where table looks like this:
A | B
---|---
MAX | MAX
MIN | MIN
MAX | MIN
MIN | MAX

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]

File: crates/proof-of-sql/src/base/database/column.rs (hunk 1 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/base/database/column.rs:244]: I just made it private. I do think the match is not going to be great since we plan to add even more integer types.

**Comment by JayWhite2357** [comment]
See https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/20 for how I think we should have the tests look going forward.

File: crates/proof-of-sql/src/sql/ast/provable_expr_plan.rs [hunk 4]
```diff
@@ -34,6 +35,10 @@ pub enum ProvableExprPlan<C: Commitment> {
     Equals(EqualsExpr<C>),
     /// Provable AST expression for an inequality expression
     Inequality(InequalityExpr<C>),
+    /// Provable numeric `+` expression
+    Add(AddSubtractExpr<C>),
+    /// Provable numeric `-` expression
+    Subtract(AddSubtractExpr<C>),
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/ast/provable_expr_plan.rs:41]: NIT:
I don't think this needs to be separate variants.
If you want separate variants, I would suggest making `is_subtract` a const generic instead and then putting the appropriate const in the variant.
As it is, it's possible to have an Add variant that acts as a Subtract.

**Review by JayWhite2357 — APPROVED** [review]
Looks good. One NIT.

**Review by JayWhite2357 — DISMISSED** [review]

**Review by JayWhite2357 — APPROVED** [review]
